src/coreclr/nativeaot/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -204,6 +204,14 @@ public static void MemoryBarrier()
         }
         #endregion
 
+        #region ReadMemoryBarrier
+        [Intrinsic]
+        internal static void ReadMemoryBarrier()
```

Comment:
`Intrinsic` methods can be implemented by recursive expansion like this:

```
[Intrinsic]
internal static void ReadMemoryBarrier();
```

It saves you from all plumbing for FCalls, implementing it in the unmanaged PAL. Also, it guarantees that the intrinsic expansion matches the non-intrinsic implementation. 

It would be better to use this here. I think you can redo both MemoryBarrier and ReadMemoryBarrier, for both naot and coreclr like this.

Reply:
For reference, this is the logic in the JIT that takes care of it: https://github.com/dotnet/runtime/blob/5b8437d9f9e410bf672bb62d5371fdeb8931e5f1/src/coreclr/jit/importercalls.cpp#L2303-L2304

Reply:
I tried putting failing implementations and it looks like Debug can reach them. Are we doing intrinsic substitution differently in Debug?

Reply:
Ah, I see - if recursive, must always do intrinsic.

Reply:
Also, you should be able to move these to src\libraries\System.Private.CoreLib\src\System\Threading\Interlocked.cs. It would be nice to add a test that creates a delegate pointing to MemoryBarrier, and calls the delegate - to make sure that the recursive intrinsic expansion is kicking in.

Reply:
> I think you can redo both MemoryBarrier and ReadMemoryBarrier, for both naot and coreclr like this.

And nonobject CompareExchange, I suppose. 
I remember I did not like something about the asm implementation. It is hard to make it right - make sure it is a full fence, uses 8.1 atomics where available. It’d be better to just leave it to the compiler to emit the best.

Reply:
> And nonobject CompareExchange, I suppose.

We do not have CompareExchange intrinsic expansion implemented for all platforms in RyuJIT. It would have to be under ifdefs that must be in sync with what's implemented in the JIT. I would leave it for separate PR.

Reply:
I've switched MemoryBarrier and ReadMemoryBarrier to be self-referring intrinsics. 
`PalMemoryBarrier` in NativeAOT is still needed as we call it from native code, but otherwise a lot of code was removed.